R/appMetadata.R

```diff
@@ -0,0 +1,378 @@
+        x = "No files matching {.str {ext}}."
+      ))
+    }
+  }
+
+  if (sum(matching) > 1) {
+    # if we have multiple matches, pick the first
+    appFiles[matching][[1]]
+  } else {
+    appFiles[matching]
+  }
+}
+
+appIsDocument <- function(appMode) {
+  appMode %in% c(
+    "rmd-static",
+    "rmd-shiny",
+    "quarto-static",
+    "quarto-shiny"
+  )
+}
+
+appHasParameters <- function(appDir, appPrimaryDoc, appMode, contentCategory = NULL) {
+  # Only Rmd deployments are marked as having parameters. Shiny applications
+  # may distribute an Rmd alongside app.R, but that does not cause the
+  # deployment to be considered parameterized.
+  #
+  # https://github.com/rstudio/rsconnect/issues/246
+  if (!(appIsDocument(appMode))) {
+    return(FALSE)
+  }
+  # Sites don't ever have parameters
+  if (identical(contentCategory, "site")) {
+    return(FALSE)
+  }
+
+  # Only Rmd files have parameters.
+  if (tolower(tools::file_ext(appPrimaryDoc)) == "rmd") {
+    filename <- file.path(appDir, appPrimaryDoc)
+    yaml <- yamlFromRmd(filename)
+    if (!is.null(yaml)) {
+      params <- yaml[["params"]]
+      # We don't care about deep parameter processing, only that they exist.
+      return(!is.null(params) && length(params) > 0)
+    }
+  }
+  FALSE
+}
+
+detectPythonInDocuments <- function(appDir, files = NULL) {
+  if (is.null(files)) {
+    # for testing
+    files <- bundleFiles(appDir)
+  }
+
+  rmdFiles <- grep("^[^/\\\\]+\\.[rq]md$", files, ignore.case = TRUE, perl = TRUE, value = TRUE)
+  for (rmdFile in rmdFiles) {
+    if (documentHasPythonChunk(file.path(appDir, rmdFile))) {
+      return(TRUE)
+    }
+  }
+  return(FALSE)
+}
+documentHasPythonChunk <- function(filename) {
+  lines <- readLines(filename, warn = FALSE, encoding = "UTF-8")
+  matches <- grep("`{python", lines, fixed = TRUE)
+  return(length(matches) > 0)
+}
+
+
+# Attempt to gather Quarto version and engines, first from quarto inspect if a
+# quarto executable is provided, and then from metadata.
+inferQuartoInfo <- function(appDir, appPrimaryDoc, appFiles, quarto, metadata) {
+  quartoInfo <- NULL
+  if (!is.null(metadata$quarto_version)) {
+    # Prefer metadata, because that means someone already ran quarto inspect
+    quartoInfo <- list(
+      "version" = metadata[["quarto_version"]],
+      "engines" = metadata[["quarto_engines"]]
+    )
+  }
+  if (is.null(quartoInfo) && !is.null(quarto)) {
+    # If we don't yet have Quarto details, run quarto inspect ourselves
+
+    # If no appPrimaryDoc has been provided, we will use the file that will be
+    # chosen if this deployment ends up with an R Markdown or Quarto app mode.
+    # This works because:
+
+    # - App modes are only used to gate primary doc inference; the behavior does
+    #   not differ between app modes.
+    # - inferAppPrimaryDoc() returns appPrimaryDoc() if it is not null.
+
+    # TODO(HW): eliminate this code by running inferAppPrimaryDoc() before
+    # inferQuartoInfo()
```

Comment:
@toph-allen - do you still have the context behind this two-pass workflow?

Callers like the `quarto` package and the IDE may provide Quarto metadata, but that's not guaranteed.

Reply:
Yeah, the context is in [this PR](https://github.com/rstudio/rsconnect/pull/604).

If the user passes a Quarto path to the `quarto` argument, we want to deploy that app as Quarto if it's Quarto-compatible (even if it only has Rmds, not qmds.).

This leads to a silly circular dependency. This is a little hazy, because it was 8 months ago, but something like: 

- If the user hasn't provided a primary doc, we need to infer it, e.g. `deployApp(appDir = ".", quarto = "quarto")` where `"."` is not a Quarto project but a dir containing an Rmd.
- Primary doc inference depends on app mode.
- App mode inference depends on the result of `quarto inspect`.
- If a user doesn't explicitly pass a primary doc, we still want a primary doc to `quarto inspect`.

I can try to explain it in a different way.

I need to get offline now. Will review the full PR in the morning.

Reply:
I think what's not obvious from that PR (because the functions were previously quite spread out), is that we could just call `inferAppPrimaryDoc()` first. I'm pretty sure this is a safe change, but we can discuss it more in the next PR.